File: src/ios/ComuOsamPlugin.swift
import OSAMCommon

@objc(ComuOsamPlugin)
class ComuOsamPlugin : CDVPlugin {
  var osamCommons : OSAMCommons?


  @objc(initialize:)
  func initialize(command: CDVInvokedUrlCommand) {
    
    guard let isDev = command.argument(at: 0) as? Bool else {
      let pluginResult = CDVPluginResult(status: .error, messageAs: "Invalid parameter")
      self.commandDelegate?.send(pluginResult, callbackId: command.callbackId)
      return
    }

    let backendURL = isDev
      ? "https://dev-osam-modul-comu.dtibcn.cat/"
      : "https://osam-modul-comu.dtibcn.cat/"

    self.osamCommons = OSAMCommons(
      vc: self.viewController,
      backendEndpoint: backendURL,
      crashlyticsWrapper: CrashlyticsWrapperIOS(),
      performanceWrapper: PerformanceWrapperIOS(),
      analyticsWrapper: AnalyticsWrapperIOS(),
      platformUtil: PlatformUtilIOS()
    )

    let pluginResult = CDVPluginResult(status: .ok, messageAs: "Configured with isDev=\(isDev)")
    self.commandDelegate?.send(pluginResult, callbackId: command.callbackId)    

  }


  /*
  // Retorna l'idioma que s'ha de passar a la crida a la llibreria
  */
  func getLanguage(params: [String:Any]) -> Language {
    switch (params["language"] as! String) {
      case "ca":
        return Language.ca
      case "es":
        return Language.es
      case "en":
        return Language.en
      default:
        return Language.ca
    }

  }

  // funció per mostrar un diàleg per actualitzar l'app si és necessari
  @objc(versionControl:)
  func versionControl(command: CDVInvokedUrlCommand) {

    guard let osamCommons = self.osamCommons else {
      let pluginResult = CDVPluginResult(status: .error, messageAs: "OSAMCommons not configured (Version Control)")
      self.commandDelegate?.send(pluginResult, callbackId: command.callbackId)
      return
    }

    // obtenim els paràmetres
    let params = command.arguments[0] as! [String:Any]

    // establim l'idioma
    var language = getLanguage(params: params)

    do {
        osamCommons.versionControl(
            language: language,
            f: { versionControlResponse in
                let responseString = String(versionControlResponse.name);
                // establim el resultat de plugin a success
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: responseString)
                print(versionControlResponse.name)
                //Envime  el resultat positiu
                self.commandDelegate!.send(pluginResult, callbackId: command.callbackId)
            }
        )
    } catch _ {
      // enviem un resultat d'error al callback
      let pluginResult = CDVPluginResult (status: CDVCommandStatus_ERROR, messageAs: "The plugin failed")
      // enviem un error al callback
      self.commandDelegate!.send(pluginResult, callbackId: command.callbackId)
    }
  }

  // funció per mostrar un diàleg per valorar l'app
  @objc(rating:)
  func rating(command: CDVInvokedUrlCommand) {

    guard let osamCommons = self.osamCommons else {
      let pluginResult = CDVPluginResult(status: .error, messageAs: "OSAMCommons not configured (Rating)")
      self.commandDelegate?.send(pluginResult, callbackId: command.callbackId)
      return
    }    

    // obtenim els paràmetres
    let params = command.arguments[0] as! [String:Any]

    // establim l'idioma
    var language = getLanguage(params: params)

    //SKStoreReviewController.requestReview()

    do {
        osamCommons.rating(
            language: language,
            f: { ratingControlResponse in

                let responseString = String(ratingControlResponse.name)
                // establim el resultat de plugin a success
                var pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs:responseString)
                print(ratingControlResponse.name)
                self.commandDelegate!.send(pluginResult, callbackId: command.callbackId)
            }
        )
    } catch _  {
      // enviem un resultat d'error al callback
      let pluginResult = CDVPluginResult (status: CDVCommandStatus_ERROR, messageAs: "The plugin failed")
      // enviem un error positiu al callback
      self.commandDelegate!.send(pluginResult, callbackId: command.callbackId)
    }
  }
}

File: src/android/ComuOsamPlugin.java
package com.comuosam.cordova.plugin;

import android.content.Context;
import android.content.Intent;
import android.util.Log;


import cat.bcn.commonmodule.ui.versioncontrol.OSAMCommons;
import cat.bcn.commonmodule.ui.versioncontrol.Language;
import cat.bcn.commonmodule.ui.versioncontrol.VersionControlResponse;
import cat.bcn.commonmodule.ui.versioncontrol.RatingControlResponse;
import cat.bcn.commonmodule.ui.versioncontrol.DeviceInformationResponse;
import cat.bcn.commonmodule.ui.versioncontrol.AppInformationResponse;

import cat.bcn.commonmodule.platform.PlatformUtilAndroid;




import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CordovaWebView;
import org.apache.cordova.PluginResult;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import kotlin.Unit;
import kotlin.jvm.functions.Function1;

import android.widget.Toast;


public class ComuOsamPlugin extends CordovaPlugin {
    OSAMCommons osamCommons;
     
    @Override
    public void initialize(CordovaInterface cordova, CordovaWebView webView) {
        super.initialize(cordova, webView);

        //Context context = cordova.getActivity().getApplicationContext();
        Context context = cordova.getActivity();
        if (context != null) {
            Log.d("osam", "Context obtained successfully.");
        } else {
            Log.e("osam", "Failed to obtain context.");
        }

        CrashlyticsAndroid crashlyticsAndroid = new CrashlyticsAndroid();
        PerformanceAndroid performanceAndroid = new PerformanceAndroid();
        AnalyticsAndroid analyticsAndroid = new AnalyticsAndroid(context);
        PlatformUtilAndroid platformUtilAndroid = new PlatformUtilAndroid(context);
        
        osamCommons = new OSAMCommons(
            cordova.getActivity(),
            context,
            "https://dev-osam-modul-comu.dtibcn.cat/",
            crashlyticsAndroid,
            performanceAndroid,
            analyticsAndroid,
            platformUtilAndroid
        );
    }

    private Unit onVersionControlResponseRecieved(VersionControlResponse response, CallbackContext callbackContext) {
        Log.d("osam", "onVersionControlResponseRecieved " + response.toString());
        PluginResult pluginResult = new PluginResult(PluginResult.Status.OK, response.toString());
        callbackContext.sendPluginResult(pluginResult);
        return Unit.INSTANCE;
    }

    private Unit onRatingControlResponseRecieved(RatingControlResponse response, CallbackContext callbackContext) {
        Log.d("osam", "onRatingControlResponseRecieved " + response.toString());
        PluginResult pluginResult = new PluginResult(PluginResult.Status.OK, response.toString());
        callbackContext.sendPluginResult(pluginResult);
        return Unit.INSTANCE;
    }

    /*
    // Retorna l'idioma que s'ha de passar a la crida a la llibreria
    */
    private Language getLanguage(JSONArray args) {
        Language language = null;
        try {
            JSONObject params = args.getJSONObject(0);
            switch (params.getString("language")) {
                case "ca":
                    language = Language.CA;
                    break;
                case "es":
                    language = Language.ES;
                    break;
                case "en":
                    language = Language.EN;
                    break;
            }
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return language;
    }

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) {
        if(action.equals("versionControl")) {
            // funció per mostrar un diàleg per actualitzar l'app si és necessari
            osamCommons.versionControl(this.getLanguage(args), response -> {
                return onVersionControlResponseRecieved(response, callbackContext);
                });
        }else if (action.equals("rating")) {
            // funció per mostrar un diàleg per valorar l'app
            osamCommons.rating(this.getLanguage(args), response -> {
                return onRatingControlResponseRecieved(response, callbackContext);
                });
           
        } else if (action.equals("deviceInformation")) {
             
        } else if(action.equals("appInformation")) {
              
        } else {
            callbackContext.error("\"" + action + "\" is not a recognized action.");
            return false;            
        }

        return true;
    }
}